src/ripple/shamap/SHAMapTreeNode.h

```diff
@@ -172,11 +177,47 @@ class SHAMapInnerNode
     void updateHashDeep();
     void addRaw (Serializer&, SHANodeFormat format) const override;
     std::string getString (SHAMapNodeID const&) const override;
+    uint256 const& key() const override;
+    void invariants(bool is_v2, bool is_root = false) const override;
 
     friend std::shared_ptr<SHAMapAbstractNode>
         SHAMapAbstractNode::make(Blob const& rawNode, std::uint32_t seq,
              SHANodeFormat format, SHAMapHash const& hash, bool hashValid,
-                 beast::Journal j);
+                 beast::Journal j, SHAMapNodeID const& id);
+
+    friend class SHAMapInnerNodeV2;
+};
+
+class SHAMapTreeNode;
+
+// SHAMapInnerNodeV2 is a "version 2" inner node.  It always has at least two children
+//   unless it is the root node and there is only one leaf in the tree, in which case
+//   that leaf is a direct child of the root.
+class SHAMapInnerNodeV2
+    : public SHAMapInnerNode
```

Comment:
We are deciding at run time whether to populate with v1 or v2 nodes, and there's a lot of code that doesn't care about the distinction between v1 and v2 nodes (say just cares that it is an inner node).  So this design makes sense for this context.  But we now also have very extensive invariants tests for the shamap, testing both v1 and v2 invariants.  And one of the invariants tested is that a shamap contains only v1 or v2 inner nodes.  These invariant tests are exercised in the unittest, and can be very easily called while debugging:  `smap.invariants();`  // will assert if any (v1 or v2) invariants have not held up.